src/main/java/org/broadinstitute/hellbender/utils/hdf5/HDF5PoNCreator.java

```diff
@@ -0,0 +1,712 @@
+        return medianOfMedians;
+    }
+
+    /**
+     * Final pre-panel normalization that consists on dividing all counts by the median of
+     * its column and log it with base 2.
+     * <p>
+     *     The normalization occurs in-place.
+     * </p>
+     *
+     * @param readCounts the input counts to normalize.
+     */
+    @VisibleForTesting
+    static void normalizeAndLogReadCounts(final ReadCountCollection readCounts, final Logger logger) {
+        final RealMatrix counts = readCounts.counts();
+        final int columnCount = counts.getColumnDimension();
+        final Median medianCalculator = new Median();
+        for (int i = 0; i < columnCount; i++) {
+            final double[] colCounts = counts.getColumn(i);
+            final double median = medianCalculator.evaluate(colCounts);
+            final double invMedian = 1.0 / median;
+            for (int j = 0; j < colCounts.length; j++) {
+                colCounts[j] = Math.log(Math.max(EPSILON, colCounts[j] * invMedian)) * INV_LN_2;
+            }
+            counts.setColumn(i, colCounts);
+        }
+        logger.info("Counts normalized by the column median and changed into bits.");
+    }
+
+
+    /**
+     * Impute zero counts to the median of non-zero values in the enclosing target row.
+     *
+     * <p>The imputation is done in-place, thus the input matrix is well be modified as a result of this call.</p>
+     *
+     * @param readCounts the input and output read-count matrix.
+     */
+    @VisibleForTesting
+    static void imputeZerosCounts(final ReadCountCollection readCounts, final Logger logger) {
+
+        final RealMatrix counts = readCounts.counts();
+        final int targetCount = counts.getRowDimension();
+        final int columnCount = counts.getColumnDimension();
+
+        // Declared once outside loop to reuse across rows:
+        final int[] zeroIndexes = new int[columnCount];
+        final double[] nonZeroValues = new double[columnCount];
+        final Median medianCalculator = new Median();
+        int totalCounts = 0;
+        int totalZeroCounts = 0;
+
+        // for each row aka target:
+        for (int i = 0; i < targetCount; i++) {
+            // We classify columns as having a zero or not.
+            // we accumulate nonZero values in nonZeroValues to calculate the median
+            // where as we accumulate zero indexes that need to be updated later.
+            final double[] rowCounts = counts.getRow(i);
+            int zeroCount = 0;
+            int nonZeroCount = 0;
+            for (int j = 0; j <  rowCounts.length; j++) {
+                if (rowCounts[j] != 0.0) {
+                    nonZeroValues[nonZeroCount++] = rowCounts[j];
+                } else {
+                    zeroIndexes[zeroCount++] = j;
+                }
+            }
+            totalCounts += rowCounts.length;
+            if (zeroCount > 0) { // We only have to do anything if if there is any zeros in the row.
```

Comment:
These are not necessary changes since I have put the visit function for the RealMatrix.  Please confirm.

No action.